pennylane/interfaces/batch/__init__.py

```diff
@@ -15,13 +15,137 @@
 This subpackage defines functions for interfacing devices' batch execution
 capabilities with different machine learning libraries.
 """
-# pylint: disable=import-outside-toplevel,too-many-arguments
+# pylint: disable=import-outside-toplevel,too-many-arguments,too-many-branches
+from functools import wraps
+
+from cachetools import LRUCache
+import numpy as np
+
 import pennylane as qml
 
 from .autograd import execute as execute_autograd
 
 
-def execute(tapes, device, gradient_fn, interface="autograd", mode="best", gradient_kwargs=None):
+def cache_execute(fn, cache, pass_kwargs=False, return_tuple=True):
+    """Decorator that adds caching to a function that executes
+    multiple tapes on a device.
+
+    This decorator makes use of :attr:`.QuantumTape.hash` to identify
+    unique tapes.
+
+    - If a tape does not match a hash in the cache, then the tape
+      has not been previously executed. It is executed, and the result
+      added to the cache.
+
+    - If a tape matches a hash in the cache, then the tape has been previously
+      executed. The corresponding cached result is
+      extracted, and the tape is not passed to the execution function.
+
+    - Finally, there might be the case where one or more tapes in the current
+      set of tapes to be executed share a hash. If this is the case, duplicated
+      are removed, to avoid redundant evaluations.
+
+    Args:
+        fn (callable): The execution function to add caching to.
+            This function should have the signature ``fn(tapes, **kwargs)``,
+            and it should return ``list[tensor_like]``, with the
+            same length as the input ``tapes``.
+        cache (None or dict or Cache): The cache to use. If ``None``,

```

Comment:
```suggestion
        cache (None or dict or Cache or bool): The cache to use. If ``None``,
```
By the way, suggestions in this file are an example where an added line is automatically added and must be deleted.

Reply:
Also, what is the `Cache` case?

Reply:
Do we need `None` if we can pass a bool?

Reply:
> By the way, suggestions in this file are an example where an added line is automatically added and must be deleted.

We worked out this is because I use Windows :(

> Do we need None if we can pass a bool?

Good question. I added this because I kept getting confused and doing `cache=None` 🤔 Do you think it is overcomplicated?

> Also, what is the Cache case?

I wasn't sure what type to put here 😆 Basically, any object that can act as a cache - that is, it has `.pop()`, `get()`, and `setdefault()` methods defined. So 'dict like'?

Reply:
> Good question. I added this because I kept getting confused and doing cache=None  Do you think it is overcomplicated?

I guess this is not very user-facing code(?) so whatever works naturally should be good! Though minor preference for having the `None` _or_ `bool` option.